parlai/utils/fp16.py

```diff
@@ -0,0 +1,677 @@
+#!/usr/bin/env python3
+# Copyright (c) Facebook, Inc. and its affiliates.
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+"""
+Utility methods for mixed precision training.
+"""
+from parlai.utils.misc import warn_once
+
+from itertools import chain
+import math
+
+try:
+    import torch
+    import torch.nn.functional as F
+except ImportError:
+    raise ImportError('Parlai requires pytorch. Go to http://pytorch.org to install.')
+
+
+###################################################
+## General Utilities
+###################################################
+
+
+class FP16SafeCrossEntropy(torch.nn.Module):
+    """
+    FP16-safe cross entropy loss.
+
+    This avoids overflow in the softmax by doing the operation in FP32.
+    """
+
+    def __init__(self, ignore_index, reduction='none'):
+        super().__init__()
+        self.NULL_IDX = ignore_index
+        self.reduction = reduction
+
+    def forward(self, scores, targets):
+        return F.nll_loss(
+            F.log_softmax(scores, 1, dtype=torch.float32),
+            targets,
+            ignore_index=self.NULL_IDX,
+            reduction=self.reduction,
+        )
+
+
+def clip_grad_norm(params, max_norm):
+    """
+    Clips grad norm.
+    """
+    params = list(params)
+    if len(params) == 1:
+        p = params[0]
+        grad_norm = torch.norm(p)
+        if grad_norm > max_norm > 0:
+            clip_coef = max_norm / (grad_norm + 1e-6)
+            p.mul_(clip_coef)
+        return grad_norm
+    elif max_norm > 0:
+        return torch.nn.utils.clip_grad_norm_(params, max_norm)
+    else:
+        return torch.sqrt(
+            sum(p.grad.data.norm() ** 2 for p in params if p.grad is not None)
+        )
+
+
+def has_overflow(grad_norm):
+    """
+    Detect inf and NaN in grad_norm.
+    """
+    if grad_norm == float('inf') or grad_norm != grad_norm:
```

Comment:
woah `grad_norm != grad_norm` works for NaN?

Reply:
apparently! this was taken directly from fairseq!